sdks/python/apache_beam/testing/synthetic_pipeline.py

```diff
@@ -116,6 +118,149 @@ def process(self, element):
         yield element
 
 
+class NonLiquidShardingOffsetRangeTracker(OffsetRestrictionTracker):
+  """ A OffsetRangeTracker that doesn't allow splitting. """
+
+  def try_split(self, split_offset):
+    return  # Don't split.
+
+  def checkpoint(self):
+    return # Don't split.
+
+
+class SyntheticSDFStepRestrictionProvider(RestrictionProvider):
+  """A `RestrictionProvider` for SyntheticSDFStep.
+
+  An initial_restriction and split that operate on num_records and ignore
+  source description (element). Splits into initial_splitting_num_bundles.
+  Returns size_estimate_override as restriction size, if set. Otherwise uses
+  element size.
+
+  If initial_splitting_uneven_chunks, produces uneven chunks.
+
+  """
+
+  def __init__(self, num_records, initial_splitting_num_bundles,
+               initial_splitting_uneven_chunks, disable_liquid_sharding,
+               size_estimate_override):
+    self._num_records = num_records
+    self._initial_splitting_num_bundles = initial_splitting_num_bundles
+    self._initial_splitting_uneven_chunks = initial_splitting_uneven_chunks
+    self._disable_liquid_sharding = disable_liquid_sharding
+    self._size_estimate_override = size_estimate_override
+
+  def initial_restriction(self, element):
+    return (0, self._num_records)
+
+  def create_tracker(self, restriction):
+    if self._disable_liquid_sharding:
+      return NonLiquidShardingOffsetRangeTracker(restriction[0],
+                                                 restriction[1])
+    else:
+      # OffsetRange.new_tracker returns a RangeTracker - not RestrictionTracker.
+      return OffsetRestrictionTracker(restriction[0], restriction[1])
+
+  def split_randomly(self, restriction):
+    ''' Randomly split the restriction into the right number of bundles.'''
+    elems = restriction[1] - restriction[0]
+    bundles = self._initial_splitting_num_bundles
+    randomNums = [np.random.randint(0, elems - 1) for _ in
+                  range(0, bundles - 1)]
+    print randomNums
+    randomNums.append(0)
+    randomNums.append(elems)
+    randomNums.sort()
+    for i in range(1, len(randomNums)):
+      yield (restriction[0] + randomNums[i - 1],
+             restriction[0] + randomNums[i])
+
+  def split(self, element, restriction):
+    elems = restriction[1] - restriction[0]
```

Comment:
Can we factor out the logic here: https://github.com/apache/beam/blob/7943d5ff74df200a821d72d5ba24558f523eb322/sdks/python/apache_beam/testing/synthetic_pipeline.py#L196

and use it in both places via a shared method.

Use zipf distribution for "uneven" splits and falling back to the `_initial_splitting_num_bundles` logic for "even" size bundles.

We could either take a parameter or default to a zipf distribution with a 3.0 factor for now.

Reply:
Done. I thought about making the config name the same but initial_splitting seems less obvious then if it supports uneven chunks.

Otherwise I'll just set it at 3.0 unless we think we'll need to mess with that.

Reply:
You can also factor out the offsets_per_split logic that you have below to reuse https://github.com/apache/beam/blob/2471e3e4656806f4b1aefa2d677ef0e3256b41f1/sdks/python/apache_beam/testing/synthetic_pipeline.py#L214
I'm just trying to ensure that the two behaviors don't diverge in how the splits are computed.